start.py

```diff
@@ -0,0 +1,337 @@
+    kdc_node.put_file(KDC_KRB5_CONF_FILEPATH,
+                      re.sub(r'EXAMPLE.COM', realm,
+                             re.sub(r'example.com', cluster.network,
+                                    re.sub(r'kerberos.example.com',
+                                           r'{}.{}'.format(kdc_hostname, cluster.network),
+                                           krb5_conf_data))))
+    kdc_conf_data = kdc_node.get_file(KDC_CONF_FILEPATH)
+    kdc_node.put_file(KDC_CONF_FILEPATH,
+                      re.sub(r'EXAMPLE.COM', realm,
+                             re.sub(r'\[kdcdefaults\]',
+                                    r'[kdcdefaults]\n max_renewablelife = 7d\n max_life = 1d',
+                                    kdc_conf_data)))
+    acl_data = kdc_node.get_file(KDC_ACL_FILEPATH)
+    kdc_node.put_file(KDC_ACL_FILEPATH, re.sub(r'EXAMPLE.COM', realm, acl_data))
+
+    logger.info('Starting KDC ...')
+    kdc_commands = [
+        'kdb5_util create -s -r {realm} -P kdcadmin'.format(realm=realm),
+        'kadmin.local -q "addprinc -pw {admin_pw} admin/admin@{realm}"'.format(admin_pw='acladmin',
+                                                                               realm=realm)
+    ]
+
+    # Add the following commands before starting kadmin daemon etc.
+    if args.kerberos_principals:
+        principal_list = ['{}@{}'.format(principal, realm)
+                          for principal in args.kerberos_principals.split(',')]
+        create_principals_cmds = ['kadmin.local -q "addprinc -randkey {}"'.format(principal)
+                                  for principal in principal_list]
+        kdc_commands.extend(create_principals_cmds)
+
+        kdc_commands.append('sleep 2') # sleep few seconds to have Docker volume available
+        kdc_commands.append('rm -f {}'.format(KDC_USER_KEYTAB_FILEPATH))
+        create_keytab_cmd = 'kadmin.local -q "xst -norandkey -k {} {}" '.format(
+            KDC_USER_KEYTAB_FILEPATH, ' '.join(principal_list))
+        kdc_commands.append(create_keytab_cmd)
+
+    kdc_commands.extend([
+        'krb5kdc',
+        'kadmind',
+        'authconfig --enablekrb5 --update'
+    ])
+
+    kdc_commands.append('cp -f {} {}'.format(KDC_KRB5_CONF_FILEPATH, KERBEROS_VOLUME_DIR))
+    if args.kerberos_principals:
+        kdc_commands.append('chmod 644 {}'.format(KDC_USER_KEYTAB_FILEPATH))
+
+    kdc_node.execute(command="bash -c '{}'".format('; '.join(kdc_commands)), quiet=quiet)
+
+    logger.info('Validating service health ...')
+    _validate_kdc_health(node=kdc_node, services=['krb5kdc', 'kadmin'], quiet=quiet)
+
+    # Add DSE specific logic on KDC host
+    logger.info('Creating `dse` and `HTTP` Kerberos principals for DSE nodes ...')
+    for node in nodes:
+        key_tab_filename = '{}/{}.keytab'.format(KERBEROS_VOLUME_DIR, node.fqdn)
+        kdc_dse_commands = [
+            'rm -f {}/{}.keytab'.format(KERBEROS_VOLUME_DIR, node.fqdn),
```

Comment:
Why is this necessary?

Reply:
i think so - if an existing file (probably a left over from previous run or such) is not removed, the newly created principals will get appended and kinit using this keytab file will fail